File: bindings/swift/Sources/Zag/ZagDiscover.swift
#if os(macOS) || os(Linux)
import Foundation

/// Provider and model discovery functions for zag.
///
/// Available on macOS and Linux only (requires Foundation.Process).
public enum ZagDiscover {

    /// List all available provider names.
    ///
    /// - Parameter bin: Path to the zag binary (defaults to `ZAG_BIN` env or `"zag"`).
    /// - Returns: An array of provider name strings.
    public static func listProviders(bin: String? = nil) async throws -> [String] {
        let caps = try await getAllCapabilities(bin: bin)
        return caps.map { $0.provider }
    }

    /// Get capability declarations for a specific provider.
    ///
    /// - Parameters:
    ///   - provider: Provider name (e.g. `"claude"`, `"codex"`, `"gemini"`, `"copilot"`, `"ollama"`).
    ///   - bin: Path to the zag binary (defaults to `ZAG_BIN` env or `"zag"`).
    /// - Returns: The provider's capability declaration.
    public static func getCapability(
        provider: String,
        bin: String? = nil
    ) async throws -> ProviderCapability {
        let b = bin ?? ZagProcess.defaultBin
        return try await discoverExec(bin: b, args: ["-p", provider])
    }

    /// Get capability declarations for all providers.
    ///
    /// - Parameter bin: Path to the zag binary (defaults to `ZAG_BIN` env or `"zag"`).
    /// - Returns: An array of capability declarations, one per provider.
    public static func getAllCapabilities(
        bin: String? = nil
    ) async throws -> [ProviderCapability] {
        let b = bin ?? ZagProcess.defaultBin
        return try await discoverExec(bin: b, args: [])
    }

    /// Resolve a model alias for a given provider.
    ///
    /// Size aliases (`small`/`s`, `medium`/`m`/`default`, `large`/`l`/`max`) are
    /// resolved to the provider-specific model. Non-alias names pass through unchanged.
    ///
    /// - Parameters:
    ///   - provider: Provider name.
    ///   - model: Model name or alias to resolve.
    ///   - bin: Path to the zag binary (defaults to `ZAG_BIN` env or `"zag"`).
    /// - Returns: The resolved model information.
    public static func resolveModel(
        provider: String,
        model: String,
        bin: String? = nil
    ) async throws -> ResolvedModel {
        let b = bin ?? ZagProcess.defaultBin
        return try await discoverExec(bin: b, args: ["-p", provider, "--resolve", model])
    }

    // MARK: - Private

    private static func discoverExec<T: Decodable>(
        bin: String,
        args: [String]
    ) async throws -> T {
        let fullArgs = ["discover"] + args + ["--json"]
        let (stdout, stderr, exitCode) = try await collectOutput(bin: bin, args: fullArgs)

        if exitCode != 0 {
            let detail = stderr.isEmpty ? stdout : stderr
            throw ZagError(
                message: "zag exited with code \(exitCode): \(detail)",
                exitCode: Int(exitCode),
                stderr: stderr)
        }

        guard let data = stdout.data(using: .utf8) else {
            throw ZagError(
                message: "Failed to read zag output as UTF-8",
                exitCode: Int(exitCode),
                stderr: stderr)
        }

        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            let preview = String(stdout.prefix(200))
            throw ZagError(
                message: "Failed to parse zag JSON output: \(preview)",
                exitCode: Int(exitCode),
                stderr: stderr)
        }
    }

    private static func collectOutput(
        bin: String, args: [String]
    ) async throws -> (stdout: String, stderr: String, exitCode: Int32) {
        let (executableURL, fullArgs) = resolveCommand(bin: bin, args: args)

        let process = Process()
        process.executableURL = executableURL
        process.arguments = fullArgs

        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe
        process.standardInput = FileHandle.nullDevice

        do {
            try process.run()
        } catch {
            throw ZagError(
                message: "Failed to start '\(bin)': \(error)",
                exitCode: nil, stderr: "")
        }

        let stdoutData = stdoutPipe.fileHandleForReading.readDataToEndOfFile()
        let stderrData = stderrPipe.fileHandleForReading.readDataToEndOfFile()

        process.waitUntilExit()

        let stdout = String(data: stdoutData, encoding: .utf8) ?? ""
        let stderr = String(data: stderrData, encoding: .utf8) ?? ""

        return (stdout, stderr, process.terminationStatus)
    }

    /// Resolve binary to an executable URL and adjusted arguments.
    private static func resolveCommand(bin: String, args: [String]) -> (URL, [String]) {
        if bin.contains("/") {
            return (URL(fileURLWithPath: bin), args)
        }
        return (URL(fileURLWithPath: "/usr/bin/env"), [bin] + args)
    }
}
#endif

File: bindings/java/src/main/java/io/zag/ZagDiscover.java
package io.zag;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/** Discovery helpers for querying provider capabilities via the zag CLI. */
public final class ZagDiscover {

    private static final ObjectMapper MAPPER =
            new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

    private ZagDiscover() {}

    /**
     * Run a zag discover subcommand and return the raw stdout.
     */
    private static String discoverExec(String bin, List<String> extraArgs) throws ZagException {
        List<String> command = new ArrayList<>();
        command.add(bin);
        command.add("discover");
        command.addAll(extraArgs);
        command.add("--json");

        try {
            ProcessBuilder pb = new ProcessBuilder(command);
            pb.redirectErrorStream(false);
            Process process = pb.start();

            String stdout =
                    new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
            String stderr =
                    new String(process.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);

            int exitCode = process.waitFor();
            if (exitCode != 0) {
                String msg = stderr.isEmpty() ? stdout : stderr;
                throw new ZagException(
                        "zag exited with code " + exitCode + ": " + msg, exitCode, stderr);
            }

            return stdout;
        } catch (ZagException e) {
            throw e;
        } catch (IOException | InterruptedException e) {
            throw new ZagException("Failed to run zag: " + e.getMessage(), null, "");
        }
    }

    /**
     * List all available provider names.
     *
     * @param bin path to the zag binary (defaults to {@code ZAG_BIN} env or {@code "zag"})
     * @return list of provider name strings
     */
    public static List<String> listProviders(String bin) throws ZagException {
        List<ProviderCapability> caps = getAllCapabilities(bin);
        List<String> names = new ArrayList<>(caps.size());
        for (ProviderCapability cap : caps) {
            names.add(cap.provider());
        }
        return names;
    }

    /** List all available provider names using the default binary. */
    public static List<String> listProviders() throws ZagException {
        return listProviders(ZagProcess.defaultBin());
    }

    /**
     * Get capability declarations for a specific provider.
     *
     * @param provider provider name (e.g. "claude", "codex", "gemini", "copilot", "ollama")
     * @param bin path to the zag binary
     * @return the provider's capability declaration
     */
    public static ProviderCapability getCapability(String provider, String bin)
            throws ZagException {
        String json = discoverExec(bin, List.of("-p", provider));
        try {
            return MAPPER.readValue(json, ProviderCapability.class);
        } catch (IOException e) {
            throw new ZagException(
                    "Failed to parse zag JSON output: " + json.substring(0, Math.min(json.length(), 200)),
                    null, "");
        }
    }

    /** Get capability declarations for a specific provider using the default binary. */
    public static ProviderCapability getCapability(String provider) throws ZagException {
        return getCapability(provider, ZagProcess.defaultBin());
    }

    /**
     * Get capability declarations for all providers.
     *
     * @param bin path to the zag binary
     * @return list of all provider capability declarations
     */
    public static List<ProviderCapability> getAllCapabilities(String bin) throws ZagException {
        String json = discoverExec(bin, List.of());
        try {
            return MAPPER.readValue(json, new TypeReference<List<ProviderCapability>>() {});
        } catch (IOException e) {
            throw new ZagException(
                    "Failed to parse zag JSON output: " + json.substring(0, Math.min(json.length(), 200)),
                    null, "");
        }
    }

    /** Get capability declarations for all providers using the default binary. */
    public static List<ProviderCapability> getAllCapabilities() throws ZagException {
        return getAllCapabilities(ZagProcess.defaultBin());
    }

    /**
     * Resolve a model alias for a given provider.
     *
     * <p>Size aliases ({@code small}/{@code s}, {@code medium}/{@code m}/{@code default},
     * {@code large}/{@code l}/{@code max}) are resolved to the provider-specific model.
     * Non-alias names pass through unchanged.
     *
     * @param provider provider name
     * @param model model name or alias to resolve
     * @param bin path to the zag binary
     * @return the resolved model information
     */
    public static ResolvedModel resolveModel(String provider, String model, String bin)
            throws ZagException {
        String json = discoverExec(bin, List.of("-p", provider, "--resolve", model));
        try {
            return MAPPER.readValue(json, ResolvedModel.class);
        } catch (IOException e) {
            throw new ZagException(
                    "Failed to parse zag JSON output: " + json.substring(0, Math.min(json.length(), 200)),
                    null, "");
        }
    }

    /** Resolve a model alias for a given provider using the default binary. */
    public static ResolvedModel resolveModel(String provider, String model) throws ZagException {
        return resolveModel(provider, model, ZagProcess.defaultBin());
    }
}
